### PR TITLE
Feature/user profile

### PR DESCRIPTION
**Pull request by siddharth-vinsol**
* After verification, the user can fill his profile.
   * For now he can upload profile pic
   * Add Profile edit page with Change password

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -13,7 +13,12 @@
   <body>
     <header class="main">
       <div>Quora Clone App</div>
-      <%= button_to 'Logout', session_path, method: :delete, class: 'logout-button' if cookies.encrypted[:user_id] %>
+      <% if cookies.encrypted[:user_id] %>
```
Comment (shubhamjuneja1) [inline app/views/layouts/application.html.erb]: signed_in?

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -1 +1,9 @@
-<h1>WELCOME <%= @user.name %></h1>
\ No newline at end of file
+<h1>WELCOME <%= current_user.name %></h1>
+
+<div class="profile">
+  <%= image_tag current_user.profile_image.signed_id ? rails_blob_path(current_user.profile_image, disposition: 'attachment') : 'default.jpeg', class: 'profile-img' %>
```
Comment (shubhamjuneja1) [inline app/views/users/show.html.erb]: extract a method to get user profile picture

File: config/routes.rb [hunk 3]
```diff
@@ -10,7 +10,15 @@
   resource :registration, path: 'signup', only: [:create] do
     get '/', action: 'new'
   end
-  resource :user, only: [:show]
+
+  scope 'profile', controller: :users, as: :profile do
```
Comment (shubhamjuneja1) [inline config/routes.rb]: why scope profile?

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 3 above)
Comment (siddharth-vinsol) [reply config/routes.rb]: Remove :profile with :user paths

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -14,4 +14,6 @@ class ApplicationController < ActionController::Base
   private def current_user
     @user ||= User.find_by(id: cookies.encrypted[:user_id])
   end
+  helper_method :current_user
```
Comment (shubhamjuneja1) [inline app/controllers/application_controller.rb]: use comma separated

File: app/views/passwords/change.html.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<div class="form">
+  <div class="form-header">Update Password</div>
+  <%= form_with model: @user, url: change_password_path, method: :patch do |form| %>
+    <% if @user.errors.any? %>
+      <div class='form-error'>
+        <ul>
+          <% @user.errors.each do |error| %>
```
Comment (shubhamjuneja1) [inline app/views/passwords/change.html.erb]: use shared error partial everywhere

File: app/models/user.rb [hunk 6]
```diff
@@ -8,6 +8,7 @@ class User < ApplicationRecord
   }
 
   has_secure_password
+  has_one_attached :profile_image, dependent: :destroy
```
Comment (shubhamjuneja1) [inline app/models/user.rb:13]: add a validation in model as well to restrict file type as well

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: app/views/users/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+<div class="form">
+  <div class="form-header">Update Profile</div>
+  <%= form_with model: current_user, url: user_path do |form| %>
+    <%= render partial: 'shared/form_errors', locals: { model_object: @user } %>
+
+    <div class="img-field">
+      <%= image_tag user_profile_image_link(current_user), class: 'profile-img' %>
+      <%= form.label :profile_image, 'Upload New Profile Picture' %>
+      <%= form.file_field :profile_image, accept: 'image/png, image/jpeg' %>
```
Comment (shubhamjuneja1) [inline app/views/users/edit.html.erb]: extract 'image/png, image/jpeg' in a constant

File: app/models/user.rb [hunk 8]
```diff
@@ -8,11 +8,13 @@ class User < ApplicationRecord
...
 
   validates :name, :email, presence: true
   validates :password, :password_confirmation, presence: true, if: :setting_password?
   validates :email, uniqueness: true
   validates :email, format: { with: QuoraClone::RegexConstants::EMAIL_REGEX }, allow_blank: true
+  validates :profile_image, attached_file_type: { types: ['.png', '.jpeg'] }, allow_blank: true
```
Comment (shubhamjuneja1) [inline app/models/user.rb]: '.png', '.jpeg' constant

File: app/helpers/users_helper.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+module UsersHelper
+  def user_profile_image_link(user)
+    current_user.profile_image.present? ? rails_blob_path(current_user.profile_image, disposition: 'attachment') : 'default.jpeg'
```
Comment (shubhamjuneja1) [inline app/helpers/users_helper.rb:3]: Will 'default.jpeg' only will work? Don't you need exact path?
Is this already in repo?

File: app/models/concerns/attached_file_type_validator.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class AttachedFileTypeValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    filename = value.filename.to_s
+    record.errors.add :attribute, I18n.t('invalid_filetype', file_types: options[:types].join(' ')) unless filename.ends_with?(*options[:types])
```
Comment (shubhamjuneja1) [inline app/models/concerns/attached_file_type_validator.rb]: what if I upload a pdf with name abc.png?

File: app/helpers/users_helper.rb (hunk 9 above)
Comment (siddharth-vinsol) [reply app/helpers/users_helper.rb:3]: Yes, already in repo in assets folder
[Image](https://github.com/siddharth-vinsol/quora-clone/blob/feature/user-profile/app/assets/images/default.jpeg)

File: app/models/concerns/attached_file_type_validator.rb (hunk 10 above)
Comment (siddharth-vinsol) [reply app/models/concerns/attached_file_type_validator.rb]: validate with content-type

File: app/views/shared/_form_field_error.html.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<% if model_object.errors[attribute].present? %>
+  <div class="form-field-error">
+    <b><%= attribute %></b>
+    <span><%= model_object.errors[attribute].first %></span>
```
Comment (shubhamjuneja1) [inline app/views/shared/_form_field_error.html.erb]: what if there are multiple errors?

**Review by shubhamjuneja1 — APPROVED** [review]

**Review by shubhamjuneja1 — APPROVED** [review]

**Pull request merged**